controllers/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -67,6 +99,18 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 		return reconcile.Result{}, nil
 	}
 
+	if tier.Annotations[toolchainv1alpha1.BundledAnnotationKey] == templates.BundledWithHostOperatorAnnotationValue {
+		if !slices.Contains(r.BundledNsTemplateTierKeys, request.NamespacedName) {
+			// this tier used to be bundled with the operator but it is not anymore.
+			// if it is not used by anything, we are safe to delete it.
+			if used, err := r.isTierUsed(ctx, tier); err != nil {
+				return reconcile.Result{}, err
+			} else if !used {
+				return reconcile.Result{}, r.Client.Delete(ctx, tier)
```

Comment:
wouldn't make more sense to move deletion of the tiers into the `pkg/templates/nstemplatetiers/nstemplatetier_generator.go` file? This file is responsible for creating the tiers, so it would also make sense to call deletions there.

Reply:
I think that the role of the reconciler is to sync the cluster state with the desired state, which is exactly what we're doing, so IMHO the reconciler is the right place. If we should move this logic into the generator it would no longer be a generator, but a "syncer" and basically become a poor man's one-time-reconciler itself.

Reply:
If we do this just once at a startup, the consequence is that if a tier is not bundled anymore but is used by any space, it will not be cleared up until the next operator restart.

If we kept the logic in the controller (and added a watch on space deletion to the nstemplatetier controller) we could remove the tier without needing the restart.

Reply:
This is enough of a corner case that I don't lean either way and am fine moving the logic to the tier generation at the operator startup. I'm just pointing out the consequence.

Reply:
As mentioned in the other message, the check of the NSTemplateTier usage can be moved to a finalizer https://github.com/codeready-toolchain/host-operator/pull/1180/files#r2128994031 this means that the "generator" can call delete the NSTemplateTier as soon as it's not bundled anymore, but the finalizer will block the removal until it's not used by any Space.